File: src/main/java/com/warren/bt/Sequence.java
package com.warren.bt;

public class Sequence extends Composite {
  @Override
  public Status tick() {
    if (!hasChildren()) {
      return Status.SUCCESS;
    }

    while (currentChild < children.size()) {
      Node child = children.get(currentChild);
      Status status = child.tick();

      if (status == Status.FAILURE) {
        reset();
        return Status.FAILURE;
      }

      if (status == Status.RUNNING) {
        return Status.RUNNING;
      }

      // Continue to next child if SUCCESS
      currentChild++;
    }

    reset();
    return Status.SUCCESS;
  }
}

File: src/main/kotlin/com/warren/bt/Sequence.kt
package com.warren.bt

class Sequence : Composite("Sequence") {
    private var currentIndex = 0

    override fun tick(): Status {
        if (currentIndex >= children.size) return Status.SUCCESS
        return when (val status = children[currentIndex].tick()) {
            Status.SUCCESS -> {
                currentIndex++
                if (currentIndex >= children.size) Status.SUCCESS else Status.RUNNING
            }

            Status.RUNNING -> Status.RUNNING

            Status.FAILURE -> Status.FAILURE
        }
    }


    override fun onReset() {
        currentIndex = 0
    }
}
